tests/test_restrict_hosts.py

```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
+import pytest
+
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse
+
+import inspect
+
+
+localhost = '127.0.0.1'
+
+connect_code_template = """
+    import socket
+    import pytest
+
+    {3}
+    def {2}():
+        socket.socket().connect(('{0}', {1}))
+"""
+
+urlopen_code_template = """
+    import pytest
+    try:
+        from urllib.request import urlopen
+    except ImportError:
+        from urllib2 import urlopen
+
+    {3}
+    def {2}():
+        assert urlopen('http://{0}:{1}/').getcode() == 200
+"""
+
+
+def assert_host_blocked(result, host):
+    result.stdout.fnmatch_lines('*A test tried to use socket.socket.connect() with host "{0}"*'.format(host))
+
+
+@pytest.fixture
+def assert_connect(testdir, httpserver):
+    def assert_socket_connect(should_pass, **kwargs):
+        # get the name of the calling function
+        test_name = inspect.stack()[1][3]
+        httpserver.serve_content('ok')
+        test_url = urlparse(httpserver.url)
+
+        mark = ''
+        cli_arg = kwargs.get('cli_arg', None)
+        code_template = kwargs.get('code_template', connect_code_template)
+        mark_arg = kwargs.get('mark_arg', None)
+
+        if mark_arg and isinstance(mark_arg, str):
```

Comment:
I'm generally not a fan of if/else logic in testing utils and the condition passed into the testing function - it's fine to have multiple fixtures/methods that do the thing explicitly, so there's less confusion of whether the test logic is accurate vs the code.

Something akin to `assert socket_connect_passes` vs `assert socket_connect_fails` or the like.

You might want to examine this bit of pytest - it likely clarifies intent and may help with the actual assertions below:
https://docs.pytest.org/en/latest/parametrize.html

Reply:
I have used parametrize before, and it could have worked for this, but i made the decision to go this way. If it's a deal breaker for you, fair enough, but it does a good job of testing the code so I'm not really up for changing it tbh.